python/revng/pypeline/daemon/notification_broker/__init__.py

```diff
@@ -0,0 +1,113 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import asyncio
+import logging
+from abc import ABC, abstractmethod
+from typing import Protocol
+
+from starlette.websockets import WebSocket
+
+from revng.pypeline.storage.storage_provider import ProjectID
+
+logger = logging.getLogger(__name__)
+
+
+class Stream(Protocol):
+    """Represents a bidirectional stream, like a TCP socket or a WebSocket"""
+
+    async def read(self) -> bytes: ...
+    async def write(self, data: bytes) -> None: ...
+
+
+class WebSocketStream(Stream):
+    """Represents a WebSocket stream"""
+
+    def __init__(self, websocket: WebSocket):
+        self.websocket = websocket
+
+    async def read(self) -> bytes:
+        return await self.websocket.receive_bytes()
+
+    async def write(self, data: bytes) -> None:
+        await self.websocket.send_bytes(data)
+
+
+class NotificationSubscriber:
+    """Represents a notification subscriber with its own message queue"""
+
+    def __init__(self, project_id: ProjectID, stream: Stream):
+        self.project_id = project_id
+        self.stream = stream
+        self.message_queue: asyncio.Queue[str] = asyncio.Queue()
+        """The broker will insert messages into this queue, and the subscriber will
+        listen for them and insert them into the message queue."""
+        self.is_active = True
+        """Flag indicating whether the subscriber is active, used so the broker
+        can garbage collect inactive subscribers."""
+
+    async def send_message(self, message: str):
+        """Add a message to this subscriber's queue"""
+        if self.is_active:
+            await self.message_queue.put(message)
+
+    async def listen_for_messages(self):
+        """Listen for messages in the queue and send them via WebSocket"""
+        try:
+            while self.is_active:
+                # Wait for a message in the queue
+                message = await self.message_queue.get()
+                if not self.is_active:
+                    break
+
+                try:
+                    await self.stream.write(message.encode())
+                except Exception as e:
+                    logger.error(
+                        f"Error sending message to stream for project {self.project_id}: {e}"
+                    )
+                    break
+        except Exception as e:
```

Comment:
Catching `Exception` is illegal. What specific exceptions do you expect to have here?